docs/broker/filtering.md

```diff
@@ -0,0 +1,360 @@
+
+Additional properties of CEL that make it attractive for Trigger Filtering
+include:
+
+*   **Familiar expression syntax.** Filter expressions written in CEL look and
+    behave like conditional expressions written in the most popular languages.
+*   **Expressions can be compiled and cached for efficiency.** We expect users
+    will create many filters that change rarely in comparison to frequency of
+    evaluations. This usage pattern works in favor of an expression language
+    that can be cached in compiled form.
+*   **Expressions are serializable and composable.** These properties are useful
+    in implementing upstream filter propagation. For example, a Broker could
+    advertise a single CEL expression composed of the union of its Triggers'
+    filters, allowing upstream sources to filter events before publishing to the
+    Broker.
+
+CEL is
+[memory-safe, side-effect-free, and terminating](https://github.com/google/cel-spec/blob/master/doc/langdef.md#overview),
+making it safe and secure to embed in a multi-tenant dispatcher process.
+
+## Examples
+
+### Single attribute exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+```
+
+Specified with a CEL expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.type == "com.github.pull.create"
+```
+
+### Multiple attributes exact match
+
+Specified with the SourceAndType field:
+
+```yaml
+spec:
+  filter:
+    sourceAndType:
+      type: com.github.pull.create
+      source: https://github.com/knative/eventing/pulls/123
+```
+
+Specified with a CEL expression:
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" &&
+        ce.source == "https://github.com/knative/eventing/pulls/123"
+```
+
+### Prefix match on source
+
+Cannot be specified with the SourceAndType field.
+
+CEL doesn't support prefix match directly, but prefix match can be accomplished
+with a regular expression:
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: ce.source.matches("^https://github.com")
+```
+
+### Complex boolean expression
+
+Cannot be specified with the SourceAndType field.
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      expression: >
+        ce.type == "com.github.pull.create" ||
+        (ce.type == "com.github.issue.create" && ce.source.matches("proposals")
+
+```
+
+### Parsed extensions
+
+Cannot be specified with the SourceAndType field.
+
+_This example assumes the repository name is available as a CloudEvents
+extension with name `repository`._
+
+_The `>` syntax is a standard yaml multiline string._
+
+```yaml
+spec:
+  filter:
+    cel:
+      parseExtensions: true
+      expression: >
+        ce.type == "com.github.pull.create" ||
+         (ce.type == "com.github.issue.create" &&
```

Comment:
When the fields and comparisons are well-known they can be split apart into a data-driven format with the option to be extended via CEL. I can see from the proposal that this is basically what you're aiming for, but the following might line up more closely with conventions in K8s and Istio.

```yaml
filter:
   matchTypes:
      - "com.github.pull.create"
      - "com.github.issue.create"
   matchSource: "https://github.com/*"
```

A trivial transformation of the data format to an expression would look as follows, and this relationship should be easy for users to follow:

```yaml
filter:
   expression: >
       (match(ce.type, "com.github.pull.create") || 
       match(ce.type, "com.github.issue.create")) && 
       match(ce.source, ["https://github.com/*"]) &&
       // added condition ...
```

A more efficient way to write this filter would be as follows:

```yaml
filter:
   expression: >
       ce.type in ["com.github.pull.create", "com.github.issue.create"] && 
       ce.source.startsWith("https://github.com") &&
       // added condition ...
```

It's also possible to combine the approaches, but when this happens the data-driven fields should be separate from the attributes available to the expression (I'm assuming filters are ORed together here):

```yaml
- filter:
   matchTypes: "com.github.pull.create"
   matchSource: "https://github.com/*"
- filter:
   matchTypes: "com.github.issue.create"
   matchSource: "https://github.com/*"
   expression: ce.repository == "proposals"
```

The model presented above is akin to the model adopted with security policies, which is that the expression is in addition to the core behavior and not an alternative to it. It's worked out pretty well in practice. Simple things are simple, and complex things are slightly less complex.